Panic during validation (can crash the server)

We recently added protection against serializing internal types (api.Taint will panic) to prevent incorrect serialization (which has been a serious problem).

However, it turns out validation errors are calling json.Marshal on values, which includes taint values passed to server validation.  @thockin you added this for readability.  Can you suggest an alternative?  As long as validation is using internal types it runs square against the safety net to prevent (rampant) use of internal objects.

```
E0124 23:14:03.726067       5 runtime.go:64] Observed a panic: "do not marshal internal struct" (do not marshal internal struct)
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:70
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:63
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:49
/usr/local/go/src/runtime/asm_amd64.s:479
/usr/local/go/src/runtime/panic.go:458
/usr/local/go/src/encoding/json/encode.go:275
/usr/local/go/src/runtime/asm_amd64.s:479
/usr/local/go/src/runtime/panic.go:458
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/api/json.go:27
<autogenerated>:46
/usr/local/go/src/encoding/json/encode.go:431
/usr/local/go/src/encoding/json/encode.go:307
/usr/local/go/src/encoding/json/encode.go:280
/usr/local/go/src/encoding/json/encode.go:145
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/validation/field/errors.go:53
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/api/errors/errors.go:186
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/genericapiserver/registry/rest/update.go:114
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/genericapiserver/registry/generic/registry/store.go:451
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/storage/etcd/etcd_helper.go:465
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/storage/cacher.go:525
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/genericapiserver/registry/generic/registry/store.go:466
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/genericapiserver/endpoints/handlers/rest.go:769
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/genericapiserver/endpoints/handlers/rest.go:1014
/usr/local/go/src/runtime/asm_amd64.s:2086
panic: do not marshal internal struct [recovered]
	panic: do not marshal internal struct [recovered]
	panic: do not marshal internal struct

goroutine 23325 [running]:
panic(0x2702ee0, 0xc4230bff90)
	/usr/local/go/src/runtime/panic.go:500 +0x1a1
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0xc425b4df28, 0x1, 0x1)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:56 +0x126
panic(0x2702ee0, 0xc4230bff90)
	/usr/local/go/src/runtime/panic.go:458 +0x243
encoding/json.(*encodeState).marshal.func1(0xc425b4cea0)
	/usr/local/go/src/encoding/json/encode.go:275 +0x19f
panic(0x2702ee0, 0xc4230bff80)
	/usr/local/go/src/runtime/panic.go:458 +0x243
k8s.io/kubernetes/pkg/api.Taint.MarshalJSON(0xc423b60060, 0x12, 0xc4235ee4b0, 0x25, 0xc4230bf6b0, 0xa, 0xc4235ee600, 0x7f1cbfcc7a98, 0x1, 0x2b09fa0, ...)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/api/json.go:27 +0x6d
k8s.io/kubernetes/pkg/api.(*Taint).MarshalJSON(0xc4235ee6f0, 0x2b09fa0, 0xc4235ee6f0, 0xc425b4cdf0, 0xc4235ee6f0, 0x4b55ca0)
	<autogenerated>:46 +0x63
encoding/json.marshalerEncoder(0xc422590790, 0x2b09fa0, 0xc4235ee6f0, 0x99, 0xc4235e0100)
	/usr/local/go/src/encoding/json/encode.go:431 +0xb1
encoding/json.(*encodeState).reflectValue(0xc422590790, 0x2b09fa0, 0xc4235ee6f0, 0x99, 0x100)
	/usr/local/go/src/encoding/json/encode.go:307 +0x82
encoding/json.(*encodeState).marshal(0xc422590790, 0x2b09fa0, 0xc4235ee6f0, 0x100, 0x0, 0x0)
	/usr/local/go/src/encoding/json/encode.go:280 +0xb8
encoding/json.Marshal(0x2b09fa0, 0xc4235ee6f0, 0x2da860a, 0x13, 0x7f1cc0039000, 0x0, 0x0)
	/usr/local/go/src/encoding/json/encode.go:145 +0x8f
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/validation/field.(*Error).ErrorBody(0xc422eafa40, 0x0, 0x2)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/validation/field/errors.go:53 +0x94
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/api/errors.NewInvalid(0x0, 0x0, 0x24a9804, 0x4, 0xc422d42f30, 0x27, 0xc4230bff30, 0x2, 0x2, 0x2)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/api/errors/errors.go:186 +0xab
k8s.io/kubernetes/pkg/genericapiserver/registry/rest.BeforeUpdate(0x4b4f3e0, 0xc420872180, 0x7f1cbfd9d3c8, 0xc422d42e70, 0x4b20ca0, 0xc424f62a00, 0x4b20ca0, 0xc424f62000, 0x0, 0x2b16420)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/genericapiserver/registry/rest/update.go:114 +0x42c
k8s.io/kubernetes/pkg/genericapiserver/registry/generic/registry.(*Store).Update.func1(0x4b20ca0, 0xc424f62000, 0x0, 0x20d, 0xc424f62000, 0x0, 0x0, 0x1, 0xc425524a00)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/genericapiserver/registry/generic/registry/store.go:451 +0x303
k8s.io/kubernetes/pkg/storage/etcd.(*etcdHelper).GuaranteedUpdate(0xc420268000, 0x7f1cbfd9d388, 0xc422d42e70, 0xc422eaebc0, 0x39, 0x4b20ca0, 0xc4251bb400, 0x1, 0xc421876f58, 0xc42532c1c0, ...)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/storage/etcd/etcd_helper.go:465 +0x399
k8s.io/kubernetes/pkg/storage.(*Cacher).GuaranteedUpdate(0xc4207e7e10, 0x7f1cbfd9d388, 0xc422d42e70, 0xc422d437d0, 0x30, 0x4b20ca0, 0xc4251bb400, 0x1, 0xc421876f58, 0xc42532c1c0, ...)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/storage/cacher.go:525 +0x327
k8s.io/kubernetes/pkg/genericapiserver/registry/generic/registry.(*Store).Update(0xc420848690, 0x7f1cbfd9d3c8, 0xc422d42e70, 0xc4230b4f62, 0x27, 0x4b34620, 0xc422eaeac0, 0x0, 0x0, 0xc422e20600, ...)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/genericapiserver/registry/generic/registry/store.go:466 +0x3bd
k8s.io/kubernetes/pkg/genericapiserver/endpoints/handlers.UpdateResource.func1.2(0xc400000018, 0x3760820, 0xc422e20758, 0x1)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/genericapiserver/endpoints/handlers/rest.go:769 +0x11b
k8s.io/kubernetes/pkg/genericapiserver/endpoints/handlers.finishRequest.func1(0xc4242d3560, 0xc42532c150, 0xc4242d3500, 0xc4242d34a0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/genericapiserver/endpoints/handlers/rest.go:1014 +0x99
created by k8s.io/kubernetes/pkg/genericapiserver/endpoints/handlers.finishRequest
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/genericapiserver/endpoints/handlers/rest.go:1019 +0xd9
Flag --address has been deprecated, see --insecure-bind-address instead.
```

--- a/staging/src/k8s.io/apimachinery/pkg/api/validation/objectmeta_test.go
+++ b/staging/src/k8s.io/apimachinery/pkg/api/validation/objectmeta_test.go
@@ -306,7 +306,7 @@ func TestValidateObjectMetaUpdatePreventsDeletionFieldMutation(t *testing.T) {
 			Old:          metav1.ObjectMeta{Name: "test", ResourceVersion: "1"},
 			New:          metav1.ObjectMeta{Name: "test", ResourceVersion: "1", DeletionTimestamp: &now},
 			ExpectedNew:  metav1.ObjectMeta{Name: "test", ResourceVersion: "1", DeletionTimestamp: &now},
-			ExpectedErrs: []string{"field.deletionTimestamp: Invalid value: \"1970-01-01T00:16:40Z\": field is immutable; may only be changed via deletion"},
+			ExpectedErrs: []string{"field.deletionTimestamp: Invalid value: 1970-01-01 00:16:40 +0000 UTC: field is immutable; may only be changed via deletion"},
 		},
 		"invalid clear deletionTimestamp": {
 			Old:          metav1.ObjectMeta{Name: "test", ResourceVersion: "1", DeletionTimestamp: &now},

--- a/staging/src/k8s.io/apimachinery/pkg/util/validation/field/errors.go
+++ b/staging/src/k8s.io/apimachinery/pkg/util/validation/field/errors.go
@@ -17,8 +17,8 @@ limitations under the License.
 package field
 
 import (
-	"encoding/json"
 	"fmt"
+	"reflect"
 	"strings"
 
 	utilerrors "k8s.io/apimachinery/pkg/util/errors"
@@ -49,14 +49,30 @@ func (v *Error) ErrorBody() string {
 	case ErrorTypeRequired, ErrorTypeForbidden, ErrorTypeTooLong, ErrorTypeInternal:
 		s = fmt.Sprintf("%s", v.Type)
 	default:
-		var bad string
-		badBytes, err := json.Marshal(v.BadValue)
-		if err != nil {
-			bad = err.Error()
-		} else {
-			bad = string(badBytes)
+		value := v.BadValue
+		if reflect.TypeOf(value).Kind() == reflect.Ptr {
+			if reflectValue := reflect.ValueOf(value); reflectValue.IsNil() {
+				value = "null"
+			} else {
+				value = reflectValue.Elem().Interface()
+			}
 		}
-		s = fmt.Sprintf("%s: %s", v.Type, bad)
+		switch t := value.(type) {
+		case int64, int32, float64, float32, bool:
+			// use simple printer for simple types
+			s = fmt.Sprintf("%s: %v", v.Type, value)
+		case string:
+			s = fmt.Sprintf("%s: %q", v.Type, t)
+		case fmt.Stringer:
+			// anything that defines String() is better than raw struct
+			s = fmt.Sprintf("%s: %s", v.Type, t.String())
+		default:
+			// fallback to raw struct
+			// TODO: internal types have panic guards against json.Marshalling to prevent
+			// accidental use of internal types in external serialized form.  For now, use
+			// %#v, although it would be better to show a more expressive output in the future
+			s = fmt.Sprintf("%s: %#v", v.Type, value)
+		}
 	}
 	if len(v.Detail) != 0 {
 		s += fmt.Sprintf(": %s", v.Detail)

--- a/staging/src/k8s.io/apimachinery/pkg/util/validation/field/errors_test.go
+++ b/staging/src/k8s.io/apimachinery/pkg/util/validation/field/errors_test.go
@@ -90,7 +90,8 @@ func TestErrorUsefulMessage(t *testing.T) {
 	for _, part := range []string{
 		"foo", ErrorTypeInvalid.String(),
 		"Baz", "Qux", "Inner", "KV", "detail",
-		"1", "aoeu", "asdf", "Billy", "2",
+		"1", "aoeu", "Billy", "2",
+		// "asdf", TODO: reenable once we have a better nested printer
 	} {
 		if !strings.Contains(s, part) {
 			t.Errorf("error message did not contain expected part '%v'", part)

